### PR TITLE
[SPARK-46728][PYTHON] Check Pandas installation properly

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to check Pandas installation properly

### Why are the changes needed?

Checking Pandas installation is not working correctly, but raising improper exception when Pandas is not installed.

This issue occurs because the deleted Pandas was not actually deleted completely when related extension is installed (e.g. `pandas-stubs`).

### Does this PR introduce _any_ user-facing change?

No API change, but user-facing error message is now showing proper error message to guide:

**Before**
```python
>>> import pyspark.pandas
AttributeError: module 'pandas' has no attribute '__version__'
```

**After**
```python
>>> import pyspark.pandas
pyspark.errors.exceptions.base.PySparkImportError: [PACKAGE_NOT_INSTALLED] Pandas >= 1.4.4 must be installed; however, it was not found.
```

### How was this patch tested?

Manually tested

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: python/pyspark/sql/pandas/utils.py [hunk 1]
```diff
@@ -27,7 +27,11 @@ def require_minimum_pandas_version() -> None:
     try:
         import pandas
 
-        have_pandas = True
+        if hasattr(pandas, "__version__"):
+            have_pandas = True
+        else:
+            have_pandas = False
+            raised_error = None
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/utils.py:34]: It works to me without the fix:

```
>>> import pyspark.pandas
...
>>>
```

The problem is when you run the tests in PyCharm, it tries to import `pandas` under `pyspark`. I think we should find another way to work around this.

Comment (itholic) [reply python/pyspark/sql/pandas/utils.py:34]: Oh,, but this is reproduced when I run it in PySpark shell:

**No Pandas**
```
./bin/pyspark
import pyspark.pandas
# pyspark.errors.exceptions.base.PySparkImportError: [PACKAGE_NOT_INSTALLED] Pandas >= 1.4.4 must be installed; however, it was not found.
```

**Old Pandas**
```
./bin/pyspark
import pyspark.pandas
# pyspark.errors.exceptions.base.PySparkImportError: [UNSUPPORTED_PACKAGE_VERSION] Pandas >= 1.4.4 must be installed; however, your version is 1.4.0.
```

Maybe this is only my local issue? Did you happen to check if you tested after uninstalling Pandas?

Comment (HyukjinKwon) [reply python/pyspark/sql/pandas/utils.py:34]: Oh, okay I face the same issue.

Comment (HyukjinKwon) [reply python/pyspark/sql/pandas/utils.py:34]: Can we explain it in PR description, and add a comment please

Comment (itholic) [reply python/pyspark/sql/pandas/utils.py:36]: Sure. Adjusted comments.

**Review by HyukjinKwon — APPROVED** [review]

**Review by dongjoon-hyun — COMMENTED** [review]
Nice. Ya, I hit the same issue on those deleted packages.

Could you do the same things for the other packages like `PyArrow`, @itholic ?

**Comment by itholic** [comment]
> Could you do the same things for the other packages like PyArrow, @itholic ?

Sure. I just confirmed that other packages work as expected without any changes unlike Pandas (e.g. PyArrow)

```python
>>> import pyspark.pandas
pyspark.errors.exceptions.base.PySparkImportError: [PACKAGE_NOT_INSTALLED] PyArrow >= 4.0.0 must be installed; however, it was not found.
```

Do you have any other packages reproduce the same issue such as Pandas?

**Comment by HyukjinKwon** [comment]
@itholic can you actually check why this happens only in pandas though?

**Comment by HyukjinKwon** [comment]
My concern is that, this is sort of a hacky bandaid fix. It is a bit weird that we do this only for pandas without knowing what's exactly going on.

**Comment by itholic** [comment]
I roughly suspect that this happened due to the same package names in our project here and there (such as `pyspark.pandas`, `pyspark.sql.pandas`), so the namespace conflicts issue occur for some reason, but could not figure out the actual root cause right now. I can do some deeper investigation.

**Comment by HyukjinKwon** [comment]
> I roughly suspect that this happened due to the same package names in our project here and there (such as pyspark.pandas, pyspark.sql.pandas), so the namespace conflicts issue occur for some reason only in pandas, but could not figure out the actual root cause right now.

This one I know because the test fails sometimes with IDE for the reason.

> The reason why I suspect in this way is that because the path /../site-packages/pandas is only not deleted clearly with PySpark dev env when uninstalling pandas.

This one can also happen in other packages as well. If that's the case, we should also address the same thing in other packages, e.g., pandas udf and spark connect. It'd be great if we can at least googling and it only happens in pandas before merging this.

**Comment by itholic** [comment]
> It'd be great if we can at least googling and it only happens in pandas before merging this.

Yeah, I googled when I submitting this PR, but unfortunately couldn't figure out any clue. Let me have some more investigation today.

**Comment by itholic** [comment]
It seems like if there are extension packages that use parts of the package we're trying to remove, `pip uninstall` adds those dependencies to the `Would not remove` list and doesn't actually remove them.

In our case, `pip uninstall` prints out `Would not remove` list and doesn't actually remove them when `pandas-stubs` is still installed, and that's why `pandas` is not completely removed:

```
(pyspark-dev-env) haejoon.lee@NQ679Q495F spark % pip uninstall pandas
Found existing installation: pandas 2.1.4
Uninstalling pandas-2.1.4:
  Would remove:
    /Users/haejoon.lee/anaconda3/envs/pyspark-dev-env/lib/python3.9/site-packages/pandas-2.1.4.dist-info/*
    /Users/haejoon.lee/anaconda3/envs/pyspark-dev-env/lib/python3.9/site-packages/pandas/*
  Would not remove (might be manually added):
    /Users/haejoon.lee/anaconda3/envs/pyspark-dev-env/lib/python3.9/site-packages/pandas/__init__.pyi
    /Users/haejoon.lee/anaconda3/envs/pyspark-dev-env/lib/python3.9/site-packages/pandas/_config/__init__.pyi
    /Users/haejoon.lee/anaconda3/envs/pyspark-dev-env/lib/python3.9/site-packages/pandas/_config/config.pyi
    ...
    /Users/haejoon.lee/anaconda3/envs/pyspark-dev-env/lib/python3.9/site-packages/pandas/util/_tester.pyi
    /Users/haejoon.lee/anaconda3/envs/pyspark-dev-env/lib/python3.9/site-packages/pandas/util/_validators.pyi
    /Users/haejoon.lee/anaconda3/envs/pyspark-dev-env/lib/python3.9/site-packages/pandas/util/testing.pyi
```

So, to completely remove `pandas`, we need to uninstall `pandas-stubs` first and then uninstall `pandas`.

**Comment by itholic** [comment]
Updated PR description and comment accordingly.

**Comment by itholic** [comment]
On second thought, this issue seems like a corner case according to https://github.com/apache/spark/pull/44745#issuecomment-1894850179. Both `pandas` and `pandas-stubs` are specified as required packages, so I think it would be a good idea not to deal with problems caused by deleting required packages.

**Comment by itholic** [comment]
Let me close this PR for now, but please feel free to ping me if there is any other opinions!

**Comment by itholic** [comment]
Hmm... actually I just noticed that this harms dev testability for some case such as https://github.com/apache/spark/pull/44778, so I think maybe we better bandaid this?? WDYT @HyukjinKwon @dongjoon-hyun ?

**Comment by HyukjinKwon** [comment]
okay, let's go ahead.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**